### PR TITLE
[wsu] apply worker roles

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
https://github.com/openshift/windows-machine-config-operator/pull/87
works only ocp 4.2 or less. In 4.3, the nodeadmission restriction
plugin is blocking nodes to register themselves with worker
labels. So, I am moving the labelling part to WSU. The
eventual goal is to move this to a controller which
is capable of watching for the WMCB label and apply
worker labels.

xref: https://github.com/openshift/windows-machine-config-operator/pull/94

/cc @aravindhp @sebsoto @suhanime

### CODE REVIEW

File: tools/ansible/tasks/wsu/main.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 - hosts: localhost
   vars:
     project_root: "{{ playbook_dir }}/../../../.."
-    kubelet_location: "https://dl.k8s.io/v1.14.0/kubernetes-node-windows-amd64.tar.gz"
+    kubelet_location: "https://dl.k8s.io/v1.16.2/kubernetes-node-windows-amd64.tar.gz"
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/wsu/main.yaml]: @suhanime - Let me know, if you want this to be removed and you will address it in a separate PR

**Comment by ravisantoshgudimetla** [comment]
Will add an e2e in a bit

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 2]
```diff
@@ -136,3 +158,13 @@ func testFilesCopied(t *testing.T) {
 	}
 
 }
+
+// testWorkerLabelsArePresent tests if the worker labels are present on the Windows Node.
+func testWorkerLabelsArePresent(t *testing.T) {
+	client, err := getKubeClient()
+	require.NoErrorf(t, err, "error getting kubeclient: %v", err)
+	// Check if the Windows node has the required label needed.
+	winNodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: workerLabel})
```
Comment (suhanime) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Does this assume there's only 1 windows node in the cluster?

File: tools/ansible/tasks/wsu/main.yaml (hunk 1 above)
Comment (aravindhp) [reply tools/ansible/tasks/wsu/main.yaml]: This should be removed from here and be introduced in a separate PR as part of the work @suhanime is doing.

File: tools/ansible/tasks/wsu/main.yaml [hunk 3]
```diff
@@ -37,6 +37,18 @@
             src: "{{ tmp_dir.path }}/kubernetes/node/bin/kubelet.exe"
             dest: "{{ tmp_dir.path }}/kubelet.exe"
 
+        # Assuming bootstrap has been run, get all the Windows node bootstrapped via WMCB which have no worker labels.
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply tools/ansible/tasks/wsu/main.yaml]: I dont understand how this going to work before WINC-40 is merged

File: tools/ansible/tasks/wsu/main.yaml (hunk 1 above)
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: I've removed it.

File: tools/ansible/tasks/wsu/main.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: I changed this. I was initially thinking we can switch from `win` hosts to `localhost` when we're done bootstrapping, however I am following @sebsoto's pattern now of `delegate_to` which should solve the problem.

**Comment by ravisantoshgudimetla** [comment]
I'll rebase this once #93 merges

**Comment by ravisantoshgudimetla** [comment]
Rebased.

**Comment by ravisantoshgudimetla** [comment]
> I'll rebase this once #93 merges

Done.

File: tools/ansible/tasks/wsu/main.yaml [hunk 4]
```diff
@@ -99,3 +99,15 @@
...
       with_items: "{{ node_csrs.stdout_lines }}"
+
+    # Assuming bootstrap has been run, get all the Windows node bootstrapped via WMCB which have no worker labels.
+    - name: Get all the Windows node bootstrapped via WMCB without worker labels
+      delegate_to: localhost
+      shell: "oc get nodes -l 'node.openshift.io/os_id=Windows, !node-role.kubernetes.io/worker' -o name"
+      register: windows_nodes_without_worker_labels
+
+    # Apply worker label to the Windows nodes bootstrapped via WMCB
+    - name: Label all the Windows nodes without worker labels
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply tools/ansible/tasks/wsu/main.yaml]: We'll also be using the bootstrapped node's name in the future (tickets i'm working on), so it makes sense to do it here.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 5]
```diff
@@ -230,3 +236,14 @@ func testNoPendingCSRs(t *testing.T) {
 		}
 	}
 }
+
+// testWorkerLabelsArePresent tests if the worker labels are present on the Windows Node.
+func testWorkerLabelsArePresent(t *testing.T) {
+	// Check if the Windows node has the required label needed.
+	windowsNodes, err := k8sclientset.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: windowsLabel})
+	require.NoErrorf(t, err, "error while getting Windows node: %v", err)
+	assert.Equalf(t, len(windowsNodes.Items), 1, "expected 1 windows nodes to be present but found %v",
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Sorry, could you also add a `require` to the begining of the test, that makes sure there are zero windows nodes in the cluster?

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 6]
```diff
@@ -114,10 +118,16 @@ func TestWSU(t *testing.T) {
...
-
 	var err error
 	k8sclientset, err = getKubeClient(kubeconfig)
 	require.NoError(t, err)
+	existingWindowsNodes, err := k8sclientset.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: windowsLabel})
+	require.NoError(t, err)
+	// The test should fail, if there are existing Windows nodes in the cluster as we're just relying on the
+	// windowsLabel to be present on the node
+	// TODO: Label or annotate the windows node to have more determinism when running test suite.
+	require.Equalf(t, len(existingWindowsNodes.Items), 0, "expected 0 windows nodes to be present but found %v",
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: @sebsoto - Added the check here.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: I don't have a preference on one vs 1 but I am curious about the reason?

File: tools/ansible/tasks/wsu/main.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: Sure, I copied the code you gave me, it works fine. Thank you :)

**Comment by ravisantoshgudimetla** [comment]
> Your commit message says, E2e test added to WSU but commented currently. Not sure what you mean by that, so please update it.

Updated, thank you. When I created the PR, WSU test suite wasn't running, so I commented the code out. As the test suite is working fine, I am enabling that code path again now.

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sebsoto) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: its really a non-issue but: https://www.grammarly.com/blog/when-to-spell-out-numbers/

[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Ok, I am not going to change then :)

File: tools/ansible/tasks/wsu/main.yaml [hunk 7]
```diff
@@ -99,3 +99,30 @@
...
+
+    # Check if worker label is already applied to the Windows nodes bootstrapped via WMCB
+    # We should always get a single node that we've bootstrapped recently, if not skip this
+    # step
+    - name: Check if worker label is already applied to the Windows nodes bootstrapped via WMCB
+      delegate_to: localhost
+      shell: "oc get node {{ node_name.stdout_lines[0] }} -o yaml | grep -i node-role.kubernetes.io/worker"
+      ignore_errors: yes
+      register: checklabels
+      when: node_name.stdout_lines | length == 1
```
Comment (ravisantoshgudimetla) [inline tools/ansible/tasks/wsu/main.yaml]: Added code for skipping when labels already exists here.

File: tools/ansible/tasks/wsu/main.yaml [hunk 8]
```diff
@@ -109,8 +111,18 @@
       retries: 3
       delay: 5
 
+    # Check if worker label is already applied to the Windows nodes bootstrapped via WMCB
+    # We should always get a single node that we've bootstrapped recently, if not skip this
+    # step
+    - name: Check if worker label is already applied to the Windows nodes bootstrapped via WMCB
+      delegate_to: localhost
+      shell: "oc get node {{ node_name.stdout_lines[0] }} -o yaml | grep -i node-role.kubernetes.io/worker"
+      ignore_errors: yes
```
Comment (suhanime) [inline tools/ansible/tasks/wsu/main.yaml]: What errors are we ignoring?

File: tools/ansible/tasks/wsu/main.yaml [hunk 9]
```diff
@@ -109,8 +111,18 @@
       retries: 3
       delay: 5
 
+    # Check if worker label is already applied to the Windows nodes bootstrapped via WMCB
+    # We should always get a single node that we've bootstrapped recently, if not skip this
+    # step
+    - name: Check if worker label is already applied to the Windows nodes bootstrapped via WMCB
+      delegate_to: localhost
+      shell: "oc get node {{ node_name.stdout_lines[0] }} -o yaml | grep -i node-role.kubernetes.io/worker"
```
Comment (suhanime) [inline tools/ansible/tasks/wsu/main.yaml]: Why `node_name.stdout_lines[0]` despite of `when: node_name.stdout_lines | length == 1`?

File: tools/ansible/tasks/wsu/main.yaml [hunk 10]
```diff
@@ -99,3 +99,30 @@
       delegate_to: localhost
       shell: "oc adm certificate approve {{ item }}"
       with_items: "{{ node_csrs.stdout_lines }}"
+
+    # Get the bootstrapped windows node name. We're using the IP address of the Windows VM created
+    # in the inventory file and if the node has multiple internal IP's or external IP's it may not work well.
+    # TODO: Move this to a more deterministic way of identifying the node we just bootstrapped
+    - name: Get bootstrapped node name
+      delegate_to: localhost
+      shell: "oc get no -o wide |awk '/{{ inventory_hostname }}/ {print $1}'"
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/main.yaml]: `oc get node` please

File: tools/ansible/tasks/wsu/main.yaml [hunk 11]
```diff
@@ -99,3 +99,30 @@
...
+      register: node_name
+      until: node_name.stdout != ""
+      retries: 3
+      delay: 5
+
+    # Check if worker label is already applied to the Windows nodes bootstrapped via WMCB
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/main.yaml]: Is there any harm in reapplying the label?

File: tools/ansible/tasks/wsu/main.yaml (hunk 8 above)
Comment (aravindhp) [reply tools/ansible/tasks/wsu/main.yaml]: And is it safe to ignore all errors?

File: tools/ansible/tasks/wsu/test/e2e/wsu_test.go [hunk 12]
```diff
@@ -114,10 +118,16 @@ func TestWSU(t *testing.T) {
 	require.NotEmptyf(t, privateKeyPath, "KUBE_SSH_KEY_PATH environment variable not set")
 	require.NotEmptyf(t, playbookPath, "WSU_PATH environment variable not set")
 	require.NotEmptyf(t, clusterAddress, "CLUSTER_ADDR environment variable not set")
-
```
Comment (aravindhp) [inline tools/ansible/tasks/wsu/test/e2e/wsu_test.go]: Please revert

File: tools/ansible/tasks/wsu/main.yaml (hunk 11 above)
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: The `oc label` command fails, if the label is already applied

File: tools/ansible/tasks/wsu/main.yaml (hunk 9 above)
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: That's how we can access the first and only element as stdout_lines is a list

File: tools/ansible/tasks/wsu/main.yaml (hunk 8 above)
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: We'll just get one error out of the shell command when the label is not present, which we can ignore.

File: tools/ansible/tasks/wsu/main.yaml (hunk 11 above)
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: Reapplying of label fails unless the command overrides.

File: tools/ansible/tasks/wsu/main.yaml (hunk 9 above)
Comment (ravisantoshgudimetla) [reply tools/ansible/tasks/wsu/main.yaml]: How else would we access the single and only element in the list?
